src/Spago/Config/AST.hs

```diff
@@ -0,0 +1,995 @@
+            maybeRight <- updateExpr level right
+            case maybeRight of
+              Just EncounteredEmbed -> do
+                pure maybeRight
+
+              Just (VariableName _) -> do
+                pure maybeRight
+
+              Just (Updated newRight) -> do
+                pure $ Just $ Updated $ Dhall.Prefer charSet preferAnn left newRight
+
+              Nothing -> do
+                maybeLeft <- updateExpr level left
+                case maybeLeft of
+                  Just EncounteredEmbed -> do
+                    pure maybeLeft
+
+                  Just (VariableName _) -> do
+                    pure maybeLeft
+
+                  Just (Updated newLeft) -> do
+                    pure $ Just $ Updated $ Dhall.Prefer charSet preferAnn newLeft right
+
+                  Nothing -> do
+                    pure maybeLeft
+
+      -- recordExpr with field1.field2.field3 = update
+      Dhall.With recordExpr field update ->
+        case level of
+          WithinField -> do
+            case astMod of
+              -- The field we're updating is a `List Text` value, not a record.
+              InsertListText _ -> do
+                pure Nothing
+
+              -- The field we're updating is a `List Text` value, not a record.
+              SetText _ -> do
+                pure Nothing
+
+          AtRootExpression key | field == (key :| []) -> do
+            --    `{ ..., dependencies = ["old"] } with dependencies = ["package"]`
+            --  to
+            --    `{ ..., dependencies = ["old"] } with dependencies = ["package", "new"]`
+            maybeResult <- updateExpr WithinField update
+            case maybeResult of
+              Just EncounteredEmbed -> do
+                case astMod of
+                  --    `{ ..., dependencies = ["old"] } with dependencies = ./deps.dhall`
+                  --  to
+                  --    `{ ..., dependencies = ["old"] } with dependencies = ./deps.dhall # ["new"]`
+                  InsertListText additions -> do
+                    pure $ Just $ Updated $ Dhall.With recordExpr field $ updateListTextByWrappingListAppend additions update
+
+                  --    `{ ..., name = ["old"] } with name = ./deps.dhall`
+                  --  to
+                  --    `{ ..., name = ["old"] } with name = "new"`
+                  SetText t -> do
+                    pure $ Just $ Updated $ Dhall.With recordExpr field t
+
+              Just (VariableName _) -> do
+                -- This can't happen because the Dhall expression is invalid. There can't be a variable name
+                -- if there isn't a let binding.
+                --    `{ ..., dependencies = ["old"] } with dependencies = x`
+                pure maybeResult
+
+              Just (Updated newUpdate) -> do
+                pure $ Just $ Updated $ Dhall.With recordExpr field newUpdate
```

Comment:
While working on the tests, I found a bug! `newUpdate` should be called `newRecordExpr`. This was likely caused by a copy-paste error.

Reply:
Whoops! This is the wrong one.